cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -0,0 +1,1722 @@
+#include <azure/identity/client_secret_credential.hpp>
+#include <azure/identity/managed_identity_credential.hpp>
+#include <azure/storage/blobs.hpp>
+#include <azure/storage/files/datalake.hpp>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <sstream>
+#include <thread>
+#include <unordered_map>
+#include <utility>
+
+#include "arrow/util/windows_fixup.h"
+
+#include "arrow/buffer.h"
+#include "arrow/filesystem/filesystem.h"
+#include "arrow/filesystem/path_util.h"
+#include "arrow/filesystem/util_internal.h"
+#include "arrow/io/interfaces.h"
+#include "arrow/io/memory.h"
+#include "arrow/io/util_internal.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/util/async_generator.h"
+#include "arrow/util/atomic_shared_ptr.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/future.h"
+#include "arrow/util/key_value_metadata.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/optional.h"
+#include "arrow/util/task_group.h"
+#include "arrow/util/thread_pool.h"
+
+namespace arrow {
+
+using internal::Uri;
+
+namespace fs {
+
+static const char kSep = '/';
+
+// -----------------------------------------------------------------------
+// AzureOptions implementation
+
+AzureOptions::AzureOptions() {}
+
+std::string AzureOptions::GetAccountNameFromConnectionString(
+    const std::string& connectionString) {
+  std::map<std::string, std::string> connectionStringMap;
+  std::string::const_iterator cur = connectionString.begin();
+
+  while (cur != connectionString.end()) {
+    auto key_begin = cur;
+    auto key_end = std::find(cur, connectionString.end(), '=');
+    std::string key = std::string(key_begin, key_end);
+    cur = key_end;
+    if (cur != connectionString.end()) {
+      ++cur;
+    }
+    auto value_begin = cur;
+    auto value_end = std::find(cur, connectionString.end(), ';');
+    std::string value = std::string(value_begin, value_end);
+    cur = value_end;
+    if (cur != connectionString.end()) {
+      ++cur;
+    }
+    if (!key.empty() || !value.empty()) {
+      connectionStringMap[std::move(key)] = std::move(value);
+    }
+  }
+
+  auto getWithDefault = [](const std::map<std::string, std::string>& m,
+                           const std::string& key,
+                           const std::string& defaultValue = std::string()) {
+    auto ite = m.find(key);
+    return ite == m.end() ? defaultValue : ite->second;
+  };
+
+  std::string accountName = getWithDefault(connectionStringMap, "AccountName");
+  if (accountName.empty()) {
+    throw std::runtime_error("Cannot find account name in connection string.");
+  }
+  return accountName;
+}
+
+void AzureOptions::ConfigureAnonymousCredentials(const std::string& account_name) {
+  account_dfs_url = "https://" + account_name + ".dfs.core.windows.net/";
+  account_blob_url = "https://" + account_name + ".blob.core.windows.net/";
```

Comment:
If the account_name/account_key is not valid then it will fail to initialize the Azure SDK clients and will be captured as an error while calling a client endpoint.